File: additional_01_tcp_chat/TcpChatServer.java
// Additional Experiment 1: TCP Client-Server Chat - Server

import java.io.*;
import java.net.*;
import java.util.Scanner;

public class TcpChatServer {
    public static void main(String[] args) {
        final int PORT = 5000;

        try (ServerSocket server = new ServerSocket(PORT)) {
            System.out.println("Waiting for client on port " + PORT + "...");

            try (
                Socket socket = server.accept();             // Accept a client.
                BufferedReader input = new BufferedReader(
                    new InputStreamReader(socket.getInputStream())
                );
                PrintWriter output = new PrintWriter(
                    socket.getOutputStream(), true
                );
                Scanner scanner = new Scanner(System.in)
            ) {
                System.out.println("Client connected.");

                while (true) {                               // Continue chatting.
                    String message = input.readLine();       // Read client message.
                    if (message == null || message.equalsIgnoreCase("exit")) break;

                    System.out.println("Client: " + message);

                    System.out.print("Server: ");
                    String reply = scanner.nextLine();       // Read server reply.
                    output.println(reply);                   // Send reply.

                    if (reply.equalsIgnoreCase("exit")) break;
                }
            }
        } catch (IOException error) {
            System.out.println("Server error: " + error.getMessage());
        }
    }
}

File: additional_01_tcp_chat/tcp_chat_server.py
# Additional Experiment 1: TCP Client-Server Chat - Server
# Run the server first, then run the client in another terminal.

import socket

HOST = "0.0.0.0"
PORT = 5000

server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)  # Create TCP socket.
server.bind((HOST, PORT))                                   # Bind IP and port.
server.listen(1)                                            # Wait for one client.

print(f"Waiting for client on port {PORT}...")
connection, address = server.accept()                       # Accept a client connection.
print("Connected:", address)

while True:                                                  # Keep chatting until "exit".
    message = connection.recv(1024).decode()                # Receive client message.
    if not message or message.lower() == "exit":
        break
    print("Client:", message)

    reply = input("Server: ")                               # Read server reply.
    connection.sendall(reply.encode())                       # Send reply to client.
    if reply.lower() == "exit":
        break

connection.close()                                          # Close client connection.
server.close()                                               # Close server socket.
